whitepapers/0013_ccq.md

```diff
@@ -0,0 +1,350 @@
+
+```
+mainnet_query_request_000000000000|
+testnet_query_request_000000000000|
+devnet_query_request_0000000000000|
+
+```
+
+### Request Execution
+
+Once the request has been validated, the query module will submit the individual per-chain query requests to the appropriate watchers for execution. The watchers will submit the RPC calls
+to the node and return the results. Communication between the query module and the watchers is via a pair of golang channels per watcher, one inbound and one outbound. The watchers will
+use batch requests to minimize RPC overhead. For this to work effectively, the integrator should properly group RPC calls into the minimal set of per-chain queries.
+
+The query module will listen for responses for all of the per-chain queries. When all per-chain responses are received, the module will post the result to be published on the gossip network.
+If any of the responses fails or times out, the query module will retry periodically for up to one minute. If after a minute some of the per-chain queries were not successful, the query
+module will drop the request.
+
+### Publication of Responses
+
+If and only if the request is successfully processed, the guardian will publish a query response message over P2P using the `ccq_resp` topic. As noted
+previously, the other guardians will not see this message. Only the REST server and possibly third party integrators will see it.
+
+The query response contains both the initial query request and the results. The presence of the request allows the integrator to verify the response is what they are expecting.
+
+The response should be signed with the prefix `query_response_0000000000000000000|`. Note that it is not necessary to have different response prefixes for each environment because
+the responses are signed with the guardian key, which is different between the environments.
+
+### Guardian Configuration
+
+The guardian configuration for CCQ will consist of the following config parameters.
+
+- `ccqEnabled` - if set to `true` then the CCQ feature is enabled. Default is false.
+- `ccqAllowedRequesters` - comma separated list of signer public keys who are allowed to submit query requests. No default.
+- `ccqP2pPort` - local port used to bind the CCQ P2P channel, default is `8996`.
+- `ccqP2pBootstrap` - bootstrap peers for the CCQ P2P channel. No default (but auto generated in tilt).
+- `ccqAllowedPeers` - comma separated list of P2P peer IDs that are allowed to submit query requests.
+
+### No Query Persistence in the Guardian
+
+In order to reduce the storage burden on the guardian node, full responses are not persisted in the guardian. However, to facilitate de-duplication and authorization, some cross-chain query information may be committed to Wormchain.
```

Comment:
Agree this can be a future improvement, but at this time the data availability of the query responses is an explicit non-goal. @bruce-riley can make this slightly more explicit.